src/cljx/cats/monad/promise.cljx

```diff
@@ -0,0 +1,169 @@
+            [clojure.core.async.impl.dispatch :as dispatch]
+            [clojure.core.async.impl.mutex :as mutex]
+            [cats.core :refer [with-monad]]
+            [cats.protocols :as proto])
+
+  #+clj
+  (:refer-clojure :exclude [promise])
+
+  #+clj
+  (:import java.util.LinkedList
+           java.util.concurrent.locks.Lock))
+
+(declare promise-monad)
+(declare then)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Monad definition
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(deftype Promise [^LinkedList takes ^Lock mutex value state]
+  proto/Context
+  (get-context [_] promise-monad)
+  (get-value [_] @value)
+
+  Object
+  (toString [_]
+    (with-out-str
+      (print @state)))
+
+  impl/ReadPort
+  (take! [_ handler]
+    (.lock mutex)
+    (let [s @state]
+      (condp = s
+        ::resolved
+        (do
+          (.unlock mutex)
+          (dispatch/run #((impl/commit handler) @value))
+          nil)
+
+        ::unresolved
+        (do
+          (.add takes handler)
+          (.unlock mutex)
+          nil)
+
+        (do
+          (.unlock mutex)
+          nil))))
+
+  impl/WritePort
+  (put! [this val handler]
+    (.lock mutex)
+    (let [s @state]
+      (condp = s
+        ::resolved
+        (do
+          (dispatch/run #((impl/commit handler) @value))
+          (.unlock mutex)
+          (box false))
+
+        ::unresolved
+        (let [takesv (into [] takes)]
+          (.clear takes)
+          (vreset! value val)
+          (vreset! state ::resolved)
+
+          (doseq [takefn takesv]
+            (dispatch/run #((impl/commit takefn) val)))
+
+          (dispatch/run #((impl/commit handler) val))
+          (.unlock mutex)
+          (box true))
+
+        (do
+          (.unlock mutex)
+          (box false))))))
+
+(defn promise
+  "Promise constructor."
+  ([] (promise nil ::unresolved))
+  ([v] (promise v ::resolved))
+  ([v t] (Promise. (LinkedList.)
+                   (mutex/mutex)
+                   (volatile! v)
+                   (volatile! t))))
+
+(defn resolved?
+  [^Promise p]
+  (= ::resolved @(.-state p)))
+
+(defn then
+  "Chain promise."
+  [^Promise p handler]
+  (let [np (promise)]
+    (go
+      (let [res1 (<! p)
+            res2 (handler res1)]
+        (put! np res2)))
+    np))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Monad definition
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(def promise-monad
+  (reify
+    proto/Functor
+    (fmap [_ f mv]
+      (then mv f))
+
+    proto/Applicative
+    (pure [_ v]
+      (promise v))
+
+    (fapply [m af av]
+      (then af (fn [f] (then av f))))
+
+    proto/Monad
+    (mreturn [_ v]
+      (promise v))
+
+    (mbind [_ mv f]
+      (let [np (promise)]
+        (go
+          (let [res1 (<! mv)
+                res2 (f res1)]
+            (if (instance? Promise res2)
```

Comment:
Yes, Because of nature of dynamic language in composition of monad operations with mlet macro, you can start with one monad and stop with other.

In any case I need more research time for it. Maybe it wrong.